File: VideoWidget/TodayViewController.m
//
//  TodayViewController.m
//  VideoWidget
//
//  Created by 郭 輝平 on 10/29/14.
//  Copyright (c) 2014 Huiping Guo. All rights reserved.
//

#import "TodayViewController.h"
#import "RSTodayVideoTableViewModel.h"
#import "RSChannelService.h"
#import "UIImageView+Loading.h"
#import <NotificationCenter/NotificationCenter.h>

static const NSInteger kMaxCellNumber = 5;
static NSString *const kCellId = @"videoCell";

static const CGFloat kCellHeight = 60.0f;

// scheme
static NSString *const kLolTubeSchemeHost = @"loltube://";
static NSString *const kLolTubeSchemeVideoPath = @"";


@interface TodayViewController () <NCWidgetProviding, UITableViewDataSource, UITableViewDelegate>

@property(nonatomic, weak) IBOutlet UITableView *tableView;

@property(nonatomic, strong) RSTodayVideoTableViewModel *tableViewModel;

@end

@implementation TodayViewController

- (void)awakeFromNib {
    [super awakeFromNib];

    RSChannelService *channelService = [[RSChannelService alloc] init];
    self.tableViewModel = [[RSTodayVideoTableViewModel alloc] initWithChannelIds:[channelService channelIds]];
}

- (void)viewDidLoad {
    [super viewDidLoad];

}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];

    //load cache data
    __weak typeof(self) weakSelf = self;
    [self.tableViewModel updateCacheDataWithSuccess:^(BOOL hasCacheData) {
        if (hasCacheData) {
            [weakSelf p_reloadData];
        }
    }];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)widgetPerformUpdateWithCompletionHandler:(void (^)(NCUpdateResult))completionHandler {
    __weak typeof(self) weakSelf = self;
    [self.tableViewModel updateWithSuccess:^(BOOL hasNewData) {
        [weakSelf p_reloadData];
        completionHandler(hasNewData ? NCUpdateResultNewData : NCUpdateResultNoData);
    }                              failure:^(NSError *error) {
        completionHandler(NCUpdateResultFailed);
    }];
}

- (void)p_reloadData {
    [self.tableView reloadData];
    self.tableView.tableFooterView = [self p_tableViewFootView];
    [self setPreferredContentSize:self.tableView.contentSize];
}

-(UIView *)p_tableViewFootView{
    if (self.tableViewModel.items.count <= kMaxCellNumber) {
        return nil;
    }
    UILabel *moreInformationLabel = [[UILabel alloc] init];
    moreInformationLabel.frame = CGRectMake(18, 0, self.tableView.frame.size.width, kCellHeight);
    moreInformationLabel.textColor = [UIColor whiteColor];
    moreInformationLabel.font = [UIFont systemFontOfSize:16];
    moreInformationLabel.text = [NSString stringWithFormat:NSLocalizedString(@"VideoWidgetMoreVideos", @"Check %d more videos on LolTube"), self.tableViewModel.items.count - kMaxCellNumber];

    UITapGestureRecognizer *tapGestureRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(moreInformationLabelTapped)];
    [moreInformationLabel addGestureRecognizer:tapGestureRecognizer];
    [moreInformationLabel setUserInteractionEnabled:YES];
    return  moreInformationLabel;
}

-(void)moreInformationLabelTapped{
    [self.extensionContext openURL:[NSURL URLWithString:kLolTubeSchemeHost] completionHandler:nil];
}
- (UIEdgeInsets)widgetMarginInsetsForProposedMarginInsets:(UIEdgeInsets)defaultMarginInsets {
    return UIEdgeInsetsZero;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.tableViewModel.items.count > kMaxCellNumber ? kMaxCellNumber : self.tableViewModel.items.count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return kCellHeight;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kCellId forIndexPath:indexPath];
    RSVideoListTableViewCellVo *cellVo = (RSVideoListTableViewCellVo *) self.tableViewModel.items[(NSUInteger) indexPath.row];

    [cell.imageView asynLoadingImageWithUrlString:cellVo.defaultThumbnailUrl placeHolderImage:[UIImage imageNamed:@"DefaultThumbnail"]];
    cell.textLabel.text = cellVo.title;

    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:NO];

    RSVideoListTableViewCellVo *cellVo = (RSVideoListTableViewCellVo *) self.tableViewModel.items[(NSUInteger) indexPath.row];

    NSString *urlString = cellVo.videoId ? [NSString stringWithFormat:@"%@%@%@", kLolTubeSchemeHost, kLolTubeSchemeVideoPath, cellVo.videoId] : kLolTubeSchemeHost;
    [self.extensionContext openURL:[NSURL URLWithString:urlString] completionHandler:nil];

}

@end

File: VideoWidget/TodayViewController.swift
//
//  TodayViewController.swift
//  LolTube
//
//  Created by 郭 輝平 on 7/25/16.
//  Copyright © 2016 Huiping Guo. All rights reserved.
//

import Foundation
import UIKit
import NotificationCenter

let kMaxCellNumber: Int = 5

let kCellId: String = "videoCell"

let kCellHeight: CGFloat = 60.0

// scheme
let kLolTubeSchemeHost: String = "loltube://"

let kLolTubeSchemeVideoPath: String = ""

class TodayViewController:UIViewController, NCWidgetProviding, UITableViewDataSource, UITableViewDelegate {
    @IBOutlet weak var tableView: UITableView!
    var tableViewModel: TodayVideoTableViewModel!
    
    
    override func awakeFromNib() {
        super.awakeFromNib()
        let channelService: ChannelService = ChannelService()
        self.tableViewModel = TodayVideoTableViewModel(channelIds: channelService.channelIds())
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        //load cache data
        self.tableViewModel.updateCacheDataWithSuccess({[weak self](hasCacheData: Bool) -> Void in
            if hasCacheData {
                self?.p_reloadData()
            }
        })
    }
    
    func widgetPerformUpdate(completionHandler: @escaping (NCUpdateResult) -> Void) {
        self.tableViewModel.updateWithSuccess({[weak self](hasNewData: Bool) -> Void in
            self?.p_reloadData()
            completionHandler(hasNewData ? .newData : .noData)
            }, failure: {(error: NSError) -> Void in
                completionHandler(.failed)
        })
    }
    
    func p_reloadData() {
        self.tableView.reloadData()
        self.tableView.tableFooterView = self.p_tableViewFootView()
        self.preferredContentSize = self.tableView.contentSize
    }
    
    func p_tableViewFootView() -> UIView? {
        if self.tableViewModel.items?.count ?? 0 <= kMaxCellNumber {
            return nil
        }
        let moreInformationLabel: UILabel = UILabel()
        moreInformationLabel.frame = CGRect(x: 18, y: 0, width: self.tableView.frame.size.width, height: kCellHeight)
        moreInformationLabel.textColor = UIColor.white
        moreInformationLabel.font = UIFont.systemFont(ofSize: 16)
        moreInformationLabel.text = String(format: NSLocalizedString("VideoWidgetMoreVideos",comment: "Check %d more videos on LolTube"), self.tableViewModel.items?.count ?? 0 - kMaxCellNumber)
        let tapGestureRecognizer: UITapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(self.moreInformationLabelTapped))
        moreInformationLabel.addGestureRecognizer(tapGestureRecognizer)
        moreInformationLabel.isUserInteractionEnabled = true
        return moreInformationLabel
    }
    
    func moreInformationLabelTapped() {
        self.extensionContext?.open(URL(string: kLolTubeSchemeHost)!, completionHandler: nil)
    }
    
    func widgetMarginInsets(forProposedMarginInsets defaultMarginInsets: UIEdgeInsets) -> UIEdgeInsets {
        return UIEdgeInsets.zero
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.tableViewModel.items?.count ?? 0 > kMaxCellNumber ? kMaxCellNumber : self.tableViewModel.items?.count ?? 0
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return kCellHeight
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell: UITableViewCell = tableView.dequeueReusableCell(withIdentifier: kCellId, for: indexPath)
        guard let  cellVo: RSVideoListTableViewCellVo = self.tableViewModel.items?[Int(indexPath.row)]  else {
            return  cell
        }
        cell.imageView?.asynLoadingImage(withUrlString: cellVo.defaultThumbnailUrl, placeHolderImage: UIImage(named: "DefaultThumbnail")!)
        cell.textLabel?.text = cellVo.title
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: false)
        guard let cellVo = self.tableViewModel.items?[Int(indexPath.row)]  else {
            return
        }
        let urlString: String = cellVo.videoId != nil ? "\(kLolTubeSchemeHost)\(kLolTubeSchemeVideoPath)\(cellVo.videoId)" : kLolTubeSchemeHost
        self.extensionContext?.open(URL(string: urlString)!, completionHandler: nil)
    }
}
